### DISCUSSION TITLE
Two Raspberry pi questions

### DISCUSSION THREAD

**Discussion by seatrout** [docker-mailserver/docker-mailserver#2734] [Q&A] [answered]
The first one is whether anyone has got this running on a raspberry pi (in my case, a pi 4 with 2gb of memory and an ssd drive to store everything on). 

The second is whether I can run the email for two or three domains simultaneously — `mymaindomain.com` and `myotherdomain.com`, possibly even `myvacant.space`. There is an answer which seems to address this question on here but I can't say I understand it entirely.

**Comment by seatrout** [comment]
> The first one is whether anyone has got this running on a raspberry pi (in my case, a pi 4 with 2gb of memory and an ssd drive to store everything on).
> 

It looks as if this will not be a problem, though I can't test it myself without pointing the dns records of an existing domain in this direction. Which will take a little time. 

> The second is whether I can run the email for two or three domains simultaneously — `mymaindomain.com` and `myotherdomain.com`, possibly even `myvacant.space`. There is an answer which seems to address this question on here but I can't say I understand it entirely.

I *think* that this is going to be possible. There is a facility in the setup script which I assume does this. 
So `./setup.sh email add wombat@wombatwombat.biz`  will magically handle email for  `wombatwombat.biz`, 
assuming that the `wombatwombat.biz` has the right DNS record, even though the compose.yml file only has `wombat@mymaindomain.com`

But am I right that all of the outgoing mail for all of those domains uses the single domain in the original yml file? 

If this is the case, do I need only one rDNS setting from my provider for that domain, which will cover all the other domains I have made with the setup script? I hope so, because according to my ISP each IP address can only have one rDNS record, even though multiple ordinary DNS records can resolve to the same IP. 

But the _reverse proxy,_ as opposed to the _rDNS_ would be handled by my web server ...

**Reply by polarathene** [reply]
> It looks as if this will not be a problem, though I can't test it myself without pointing the dns records of an existing domain in this direction. Which will take a little time.

You can do local testing with a client system modifying their hosts, eg on Linux `/etc/hosts` can add any FQDN (domain) you like and point to an IP (or localhost) to resolve to.

That will take priority before querying public DNS, and if you choose a TLD like `.test` then these will not risk accidentally querying public DNS as it's not a valid TLD for public use.

So you could have mail server `mail.example.test` and domains `hello.test`, `world.test`.

---

> But am I right that all of the outgoing mail for all of those domains uses the single domain in the original yml file?

You use `hostname` in compose yml to define the mail server hostname, not the domains you want to manage (_although sometimes they share the same domain_).

You can have `mail.example.test` as your mailsever that resolves to the Pi. Then you can have your own domains to handle mail for `hello.test`, `world.test`, and if you like `example.test` as well. You add those to `postfix-accounts.cf` config file with `setup email add` command, and when the container starts it will process that and create the accounts.

Your DNS records then have your domains (`hello.test` and `world.test`) point MX records to `mail.example.test` (A record, not CNAME). This says where mail should be delivered when sent to those domains.

For sending, you need to authorize `mail.example.test` (_you'd need to do the same for `example.test` as well if you add that, as it's not the same FQDN as `mail.example.test`, just a common domain_). You do this with SPF record (_and later you can setup DKIM and DMARC if you like, those are not mandatory_), one for each domain you want to support, they will each have a TXT record that permits `mail.example.test` to send mail, and other mail servers you send to (such as gmail) will then check SPF records of those domains to confirm it matches your `hostname` (mail server HELO) that is sending the mail.

When sending mail, eg with Thunderbird, you can use any of your accounts `user@hello.test`, `user@world.test` and connect to the mail server `mail.example.test`. With password to authenticate, DMS will acknowledge the account is authorized to send mail, while rejecting anyone without an account to login with (_unless you have changed the default configuration to allow it_).

---

> If this is the case, do I need only one rDNS setting from my provider for that domain, which will cover all the other domains I have made with the setup script? I hope so, because according to my ISP each IP address can only have one rDNS record, even though multiple ordinary DNS records can resolve to the same IP.

AFAIK, what happens is your mail server `mail.example.test` will send the mail, and other mail server receiving may check rDNS based on the HELO (it should be `mail.example.test`), the FQDN (`mail.example.test`) is then looked up via DNS, and the IP checked for rDNS. This does not necessarily have to match `mail.example.test`, but doing a DNS lookup on it again should resolve to the same IP I think. Some services allow you to configure your rDNS record to match your FQDN.

This isn't a topic I'm 100% sure on, but it's roughly like that. The mail accounts involved in the mail don't matter as that's a separate verification (eg SPF), whereas rDNS is to establish trust in the mail server is who they claim to be.

> But the _reverse proxy,_ as opposed to the _rDNS_ would be handled by my web server ...

Mail server uses different ports from Web, that should not be the case, unless your reverse proxy is also handling those ports. Likewise, your mail server should be assigned a subdomain such as `mail`. Users often get confused here and think that `example.test` should be used if you want `user@example.test`.

- `mail.example.test` will point to the mail server service (DMS). That provides an IP to a server and then mail software / communications will happen of mail specific ports.
- `user@example.test` will lookup DNS for MX record for where to receive mail at (eg: `mail.example.test`), and SPF will be checked at `example.test` when a mailserver is sending on behalf of a domain. Both are DNS lookup, no reverse proxy involved at that point.

DMS can be configured with `hostname: mail.example.test` and by default will internally strip `mail.` for the Postfix `main.cf` config. This will make `example.test` the default mail domain, and configure `postmaster@example.test` for sending system mails (_eg: from system accounts doing scheduled tasks if something goes wrong, bounce mail notifications from other mail servers if sending went wrong, etc_). A system account such as `root` could then send as `root@example.test` or be aliased to `postmaster@example.test` for internal mail.

**Reply by seatrout** [reply]
Thank you for this very comprehensive reply. I will proceed inch by inch. Your suggestion of a "test" domain in /etc/hosts is really smart. Thanks!

**Reply by seatrout** [reply]
> For sending, you need to authorize `mail.example.test` (_you'd need to do the same for `example.test` as well if you add that, as it's not the same FQDN as `mail.example.test`, just a common domain_). You do this with SPF record (_and later you can setup DKIM and DMARC if you like, those are not mandatory_), one for each domain you want to support, they will each have a TXT record that permits `mail.example.test` to send mail, and other mail servers you send to (such as gmail) will then check SPF records of those domains to confirm it matches your `hostname` (mail server HELO) that is sending the mail.

I think I understand this. The SPF record for each domain I want to send email from has to be set up with my DNS registrar, and to point to the real-life external IP of the pi, same as the A record will point to — or maybe it just points to the A record: I'll find out when I get that far. This is a mechanism for the receiving domain to use to check I am not spoofing the sender. Without it my outgoing mail will likely be junked.

It is quite separate from the checks that my own SMTP sender makes to see that I am authorised to use it.

**Reply by polarathene** [reply, marked as answer]
If you like I shared a [rough example for offline testing](https://github.com/docker-mailserver/docker-mailserver/issues/2601#issuecomment-1135520225) a while back. You can scroll down through that message and see some DNS record examples, it demonstrates 3 DMS containers for testing relays.

Here is an example SPF record: `"v=spf1 a:mail.relay-service.test ip4:172.16.238.99 -all"`
- It allows the A domain `mail.relay-service.test` to send on behalf of this `example.test` domain (_which has a mailserver at `mail.example.test`_).
- It also allows a specific IP to do the same (_which was from a container in the same docker network, that had no domain_).

Another example is `"v=spf1 mx -all"`. Much smaller and for the `destination.test` DNS, it allows the MX records (where to receive mail to) of the same domain to send mail, which in this case is `mail.destination.test`, but those MX records could point to an external domain too. The MX record then resolves to an A record (`mail.destination.test`) to identify the server IP.

One last example is `"v=spf1 mx a:mail.relay-service.test -all"` for `origin.test` which allows itself to send mail by authorizing all `origin.test` MX records similar to the previous example, but also a specific A record `mail.relay-service.test` for that mail server to also have permission to send mail on behalf of `origin.test`.

Hope that helps explain things :)

Sometimes it's easier to setup locally and play with - which is what I did (_my own certs from local SmallStep CA container, CoreDNS for local DNS_), allowing for full offline testing with my fake `.test` domains.
